whatmp3.1
.TH WHATMP3 1 whatmp3\-VERSION
.SH NAME
whatmp3 \- transcode audio files and create torrents
.SH SYNOPSIS
.B whatmp3
.RB [ \-v ]
.RB [ \-h ]
.RB [ \-\-version ]
.RB [ \-n ]
.RB [ \-r ]
.RB [ \-c ]
.RB [ \-i ]
.RB [ \-s ]
.RB [ \-S ]
.RB [ \-D ]
.RB [ \-L ]
.RB [ \-C ]
.RB [ \-H ]
.RB [ \-w ]
.RB [ \-d ]
.RB [ \-M ]
.RB [ \-z ]
.RB [ \-t
.IR URL ]
.RB [ \-o
.IR DIR ]
.RB [ \-O
.IR DIR ]
.RB [ \-\-threads
.IR THREADS ]
.RB [ \-\-AAC ]
.RB [ \-\-V8 ]
.RB [ \-\-V0 ]
.RB [ \-\-ALAC ]
.RB [ \-\-V2 ]
.RB [ \-\-Q8 ]
.RB [ \-\-FLAC ]
.RB [ \-\-320 ]
.IR flacdir\ [ flacdir\ ... ]
.SH DESCRIPTION
whatmp3 transcodes audio files and creates torrents for them.
.P
whatmp3 takes a list of directories containing FLAC files and
a list of formats to transcode to. For each top level directory passed,
a new directory containing the transcoded audio files and optionally
a torrent (with 
.BR mktorrent )
are created.
.P
whatmp3 requires 
.BR flac ,
.BR metaflac ,
at least one kind of encoder (eg
.BR lame ,
.BR oggenc ).
.P
.BR mktorrent
and replaygain tools (eg
.BR vorbisgain )
are optionally required.
.P
whatmp3 will not overwrite any files in output directories unless passed
.BR -w
.P
whatmp3 will
.B touch
.B FAILURE
in each top level dir in which a critical error has occured while transcoding.
.P
.SH OPTIONS
.TP
.BR \-\-version
prints version information to standard output, then exits.
.TP
.BR \-v \fR,\ \fB\-\-verbose
prints extra information to screen
.TP
.BR \-n \fR,\ \fB\-\-notorrent
does not create a torrent even if a tracker announce url is provided
.TP
.BR \-r \fR,\ \fB\-\-replaygain
applies replaygain to transcoded audio files
.TP
.BR \-c \fR,\ \fB\-\-original
creates torrents for the original flac directories
.TP
.BR \-i \fR,\ \fB\-\-ignore
ignores top level directories that do not contain flac files
.TP
.BR \-s \fR,\ \fB\-\-silent
prints nothing to stdout
.TP
.BR \-S \fR,\ \fB\-\-skipgenre
skips adding the genre tag to output files
.TP
.BR \-D \fR,\ \fB\-\-nodate
does not write the date of creation to torrent files
.TP
.BR \-L \fR,\ \fB\-\-nolog
does not copy *.log files to output directories
.TP
.BR \-C \fR,\ \fB\-\-nocue
does not copy *.cue files to output directories
.TP
.BR \-H \fR,\ \fB\-\-nodots
do not copy dot/hidden files after conversion
.TP
.BR \-w \fR,\ \fB\-\-overwrite
overwrites files in output directories
.TP
.BR \-d \fR,\ \fB\-\-dither
dithers flacs to 16 bit / 44hz with
.BR sox
before transcoding
.TP
.BR \-M \fR,\ \fB\-\-nocopyother
do not copy non-audio files to output directories
.TP
.BR \-z \fR,\ \fB\-\-zeropad
zeropads tracknumber tags of transcoded files
.TP
.BI \-a\  args \fR,\ \fB\-\-additional= args
pass additional arguments to
.B mktorrent
.TP
.BI \-t\  tracker \fR,\ \fB\-\-tracker\  tracker
sets torrent tracker announce url to
.I tracker
.TP
.BI \-o\  dir \fR,\ \fB\-\-output\  dir 
places transcode file directories in
.I dir
.TP
.BI \-O\  dir \fR,\ \fB\-\-torrent\-dir\  dir
places torrent files in
.I dir
.TP
.BI \-T\  threads \fR,\ \fB\-\-threads\  threads
set number of transcoding threads to
.I threads
; default is defined by os.cpu_count()
.SH USAGE
whatmp3 requires a list of formats to transcode to (eg
.BR \-\-Q8 )
and a list of directories containing .flac files to transcode from.
.PP
If a tracker is provided with
.B \-\-tracker
.I tracker
a torrent will be created with
.BR mktorrent
.BR \-p
.BI \-a\  tracker
for each transcoded top level directory.
.PP
whatmp3 will replace the first occurence of
.BR FLAC
with the encoder tag (eg
.BR Q8 )
in the top level dirname.
.SH EXAMPLES
.PP
create directories 
.B Svartrit - I (Q8)
, etc in 
.B ~/tor
containing ogg vorbis Q8 and mp3 V0 transcodes with zeropadded tracknumbers and
replaygain applied, and torrents 
.B Svartrit - I (Q8).torrent
, etc in
.B ~/tor
with the specified announce URL:

.nf
	whatmp3 -rz -o ~/tor -t "http://my.tracker/announce" --Q8 --V0 "Svartrit - I" "Svartrit - II"
.fi
.SH CONFIGURATION
Configuration is handled by editing the source file to either add or change encoder options, and
to change defaults of command flags. Later releases may add support for a configuration file.
.SH SEE ALSO
.BR mktorrent (1),
.BR flac (1),
.BR metaflac (1),
.BR oggenc (1),
.BR lame (1),
.BR mp3gain (1),
.BR vorbisgain (1),
.BR sox (1)
.SH BUGS
whatmp3 has been known to choke on filenames containing certain characters, such as
.BR $ .


CHANGELOG
# 3.9
- convert regexes to raw strings (`r''`) to fix SyntaxWarnings
- support AIFF
- follow symlinks on supported systems

# 3.8
- do not append /announce to tracker
- add -H --nodots option to disable copying of dotfiles after conversion
- add --a --additional option for passing mktorrent arguments
- fix race condition when creating output directories
- fix aacgain file extension
- change alac extension to .m4a
- revert os.cpu_count() change, not all python3's have multiprocessing
- properly escape replaygain commands
- formatting changes

# 3.7
- beautify output
- improve error reporting (`touch dir/FAILURE')
- update documentation
- add Makefile
- add manpage
- add --silent
- add --ignore
- add --V8
- switch to os.cpu_count() and do not import multiprocessing
- switch to argparse
- add single character arguments for all options

# 3.6
- Generalise --skipgenre to work on all default encoders
- Generalise --replaygain similarly
- Use multithreading.cpu_count() to determine default thread count
- Become fully case insensitive
- Fix %-in-filename bug
- Fix flac-in-path bug
- Fix extra ='s bug
- Make --original respect --torrent-dir
- Add --noreplaygain to lame commands
- Rename --moveother to --copyother

# 3.5
- Don't remove FLAC from FLAC torrents
- Replace lowercase flac in torrent directory
- Respect --torrent-dir
- Add --skip-genre
- Add --no-date

# 3.4
- Ignore m3u files
- Add --nocue and --nolog options
- Change default output directory to cwd

# 3.3
- Fix and improve FLAC string removal
- Python 2.5 compatibility
- Restructure flac_command creation
- Fix quote bug

# 3.2
- Fix ffmpeg command (no '-o')
- Use os.environ['HOME'] instead of /home/user
- Improve threading
- Fix AAC tagging output (add quotes)

# 3.1
- Fix torrent creation bug
- Fix moveother bugs
- Add --FLAC support

# 3.0

- Fix bug with '`' characters in filenames
- Add threading support
- Add AAC support through neroAacEnc and neroAacTag
- Switch to Python
- Add replaygain support


whatmp3.py
#!/usr/bin/env python3

import argparse
import multiprocessing
import os
import re
import shutil
import sys
import threading
from fnmatch import fnmatch

VERSION = "3.9"

# DEFAULT CONFIGURATION

# Output folder unless specified
# output = os.path.join(os.environ['HOME'], "Desktop/")
output = os.getcwd()

# Separate torrent output folder (defaults to output):
torrent_dir = output

# Do you want to zeropad tracknumbers? (1 => 01, 2 => 02 ...)
zeropad = 1

# Do you want to dither FLACs to 16/44 before encoding?
dither = 0

# Specify tracker announce URL
tracker = None

# Additional arguments to mktorrent
additional = None

# Max number of threads (e.g., Dual-core = 2, Hyperthreaded Dual-core = 4)
max_threads = multiprocessing.cpu_count()

# Tags to copy (note: changing/adding to these requires changing/adding values in/to 'encoders' below)
copy_tags = ('TITLE', 'ALBUM', 'ARTIST', 'TRACKNUMBER', 'GENRE', 'COMMENT', 'DATE')

# Default encoding options
enc_opts = {
    '320':  {'enc': 'lame',       'ext': '.mp3',  'opts': '-q 0 -b 320 --ignore-tag-errors --noreplaygain'},
    'V0':   {'enc': 'lame',       'ext': '.mp3',  'opts': '-q 0 -V 0 --vbr-new --ignore-tag-errors --noreplaygain'},
    'V2':   {'enc': 'lame',       'ext': '.mp3',  'opts': '-q 0 -V 2 --vbr-new --ignore-tag-errors --noreplaygain'},
    'V8':   {'enc': 'lame',       'ext': '.mp3',  'opts': '-q 0 -V 8 --vbr-new --ignore-tag-errors --noreplaygain'},
    'Q8':   {'enc': 'oggenc',     'ext': '.ogg',  'opts': '-q 8 --utf8'},
    'AAC':  {'enc': 'neroAacEnc', 'ext': '.aac',  'opts': '-br 320000'},
    'ALAC': {'enc': 'ffmpeg',     'ext': '.m4a',  'opts': ''},
    'FLAC': {'enc': 'flac',       'ext': '.flac', 'opts': '--best'},
    'AIFF': {'enc': 'flac',       'ext': '.aiff', 'opts': '--best'}
}

encoders = {
    'lame': {
        'enc':         "lame --silent %(opts)s %(tags)s --add-id3v2 - '%(filename)s' 2>&1",
        'TITLE':       "--tt '%(TITLE)s'",
        'ALBUM':       "--tl '%(ALBUM)s'",
        'ARTIST':      "--ta '%(ARTIST)s'",
        'TRACKNUMBER': "--tn '%(TRACKNUMBER)s'",
        'GENRE':       "--tg '%(GENRE)s'",
        'DATE':        "--ty '%(DATE)s'",
        'COMMENT':     "--tc '%(COMMENT)s'",
        'regain':      "mp3gain -q -c -s i '%s'/*.mp3"
    },
    'oggenc': {
        'enc':         "oggenc -Q %(opts)s %(tags)s -o '%(filename)s' - 2>&1",
        'TITLE':       "-t '%(TITLE)s'",
        'ALBUM':       "-l '%(ALBUM)s'",
        'ARTIST':      "-a '%(ARTIST)s'",
        'TRACKNUMBER': "-N '%(TRACKNUMBER)s'",
        'GENRE':       "-G '%(GENRE)s'",
        'DATE':        "-d '%(DATE)s'",
        'COMMENT':     "-c 'comment=%(COMMENT)s'",
        'regain':      "vorbisgain -qafrs '%s'/*.ogg"
    },
    'neroAacEnc': {
        'enc':         "neroAacEnc %(opts)s -if - -of '%(filename)s' 2>&1 && neroAacTag %(tags)s",
        'TITLE':       "-meta:title='%(TITLE)s'",
        'ALBUM':       "-meta:album='%(ALBUM)s'",
        'ARTIST':      "-meta:artist='%(ARTIST)s'",
        'TRACKNUMBER': "-meta:track='%(TRACKNUMBER)s'",
        'GENRE':       "-meta:genre='%(GENRE)s'",
        'DATE':        "-meta:year='%(DATE)s'",
        'COMMENT':     "-meta:comment='%(COMMENT)s'",
        'regain':      "aacgain -q -c '%s'/*.aac"
    },
    'ffmpeg': {
        'enc':         "ffmpeg %(opts)s -i - -acodec alac %(tags)s '%(filename)s' 2>&1",
        'TITLE':       "-metadata title='%(TITLE)s'",
        'ALBUM':       "-metadata album='%(ALBUM)s'",
        'ARTIST':      "-metadata author='%(ARTIST)s'",
        'TRACKNUMBER': "-metadata track='%(TRACKNUMBER)s'",
        'GENRE':       "-metadata genre='%(GENRE)s'",
        'DATE':        "-metadata date='%(DATE)s'",
        'COMMENT':     "-metadata comment='%(COMMENT)s'",
        'regain':      ""
    },
    'flac': {
        'enc':         "flac %(opts)s -s %(tags)s -o '%(filename)s' - 2>&1",
        'TITLE':       "-T 'TITLE=%(TITLE)s'",
        'ALBUM':       "-T 'ALBUM=%(ALBUM)s'",
        'ARTIST':      "-T 'ARTIST=%(ARTIST)s'",
        'TRACKNUMBER': "-T 'TRACKNUMBER=%(TRACKNUMBER)s'",
        'GENRE':       "-T 'GENRE=%(GENRE)s'",
        'DATE':        "-T 'DATE=%(DATE)s'",
        'COMMENT':     "-T 'COMMENT=%(COMMENT)s'",
        'regain':      "metaflac --add-replay-gain '%s'/*.flac"
    }
}

dither_cmd = 'sox -t wav - -b 16 -t wav - rate 44100 dither'

# END CONFIGURATION

codecs = []

def copy_other(opts, flacdir, outdir):
    if opts.verbose:
        print('COPYING other files')
    for dirpath, dirs, files in os.walk(flacdir, topdown=False, followlinks=True):
        for name in files:
            if opts.nolog and fnmatch(name.lower(), '*.log'):
                continue
            if opts.nocue and fnmatch(name.lower(), '*.cue'):
                continue
            if opts.nodots and fnmatch(name.lower(), '^.'):
                continue
            if (not fnmatch(name.lower(), '*.flac')
               and not fnmatch(name.lower(), '*.m3u')):
                d = re.sub(re.escape(flacdir), outdir, dirpath)
                if (os.path.exists(os.path.join(d, name))
                   and not opts.overwrite):
                    continue
                if not os.path.exists(d):
                    os.makedirs(d)
                shutil.copy(os.path.join(dirpath, name), d)

class EncoderArg(argparse.Action):
    def __init__(self, option_strings, dest, nargs=None, **kwargs):
        super(EncoderArg, self).__init__(option_strings, dest, nargs, **kwargs)
    def __call__(self, parser, namespace, values, option_string=None):
        codecs.append(option_string[2:])

def escape_quote(pattern):
    pattern = re.sub("'", "'\"'\"'", pattern)
    return pattern

def escape_percent(pattern):
    pattern = re.sub('%', '%%', pattern)
    return pattern

def failure(r, msg):
    print("ERROR: %s: %s" % (r, msg), file=sys.stderr)

def make_torrent(opts, target):
    if opts.verbose:
        print('MAKE: %s.torrent' % os.path.relpath(target))
    torrent_cmd = "mktorrent -p -a '%s' -o '%s.torrent' '%s' 2>&1" % (
        opts.tracker, escape_quote(os.path.join(opts.torrent_dir,
                                   os.path.basename(target))),
        escape_quote(target)
    )
    if opts.additional:
        torrent_cmd += ' ' + opts.additional
    if opts.nodate:
        torrent_cmd += ' -d'
    if not opts.verbose:
        torrent_cmd += ' >/dev/null'
    if opts.verbose:
        print(torrent_cmd)
    r = system(torrent_cmd)
    if r: failure(r, torrent_cmd)

def replaygain(opts, codec, outdir):
    if opts.verbose:
        print("APPLYING replaygain")
        print(encoders[enc_opts[codec]['enc']]['regain'] % outdir)
    r = system(encoders[enc_opts[codec]['enc']]['regain'] % escape_quote(outdir))
    if r: failure(r, "replaygain")
    for dirpath, dirs, files in os.walk(outdir, topdown=False, followlinks=True):
        for name in dirs:
            r = system(encoders[enc_opts[codec]['enc']]['regain']
                       % os.path.join(dirpath, name))
            if r: failure(r, "replaygain")

def setup_parser():
    p = argparse.ArgumentParser(
        description="whatmp3 transcodes audio files and creates torrents for them",
        argument_default=False,
        epilog="""depends on flac, metaflac, mktorrent, and optionally oggenc, lame, neroAacEnc,
        neroAacTag, mp3gain, aacgain, vorbisgain, and sox""")
    p.add_argument('--version', action='version', version='%(prog)s ' + VERSION)
    for a in [
        [['-v', '--verbose'],     False,   'increase verbosity'],
        [['-n', '--notorrent'],   False,   'do not create a torrent after conversion'],
        [['-r', '--replaygain'],  False,   'add ReplayGain to new files'],
        [['-c', '--original'],    False,   'create a torrent for the original FLAC'],
        [['-i', '--ignore'],      False,   'ignore top level directories without flacs'],
        [['-s', '--silent'],      False,   'do not write to stdout'],
        [['-S', '--skipgenre'],   False,   'do not insert a genre tag in MP3 files'],
        [['-D', '--nodate'],      False,   'do not write the creation date to the .torrent file'],
        [['-L', '--nolog'],       False,   'do not copy log files after conversion'],
        [['-C', '--nocue'],       False,   'do not copy cue files after conversion'],
        [['-H', '--nodots'],      False,   'do not copy dot/hidden files after conversion'],
        [['-w', '--overwrite'],   False,   'overwrite files in output dir'],
        [['-d', '--dither'],      dither,  'dither FLACs to 16/44 before encoding'],
        [['-M', '--nocopyother'], False,   'do not copy additional files'],
        [['-z', '--zeropad'],     zeropad, 'zeropad tracknumbers (def: true)'],
    ]:
        p.add_argument(*a[0], **{'default': a[1], 'action': 'store_true', 'help': a[2]})
    for a in [
        [['-a', '--additional'],  additional,  'ARGS', 'additional arguments to mktorrent'],
        [['-t', '--tracker'],     tracker,     'URL',  'tracker URL'],
        [['-o', '--output'],      output,      'DIR',  'set output dir'],
        [['-O', '--torrent-dir'], torrent_dir, 'DIR',  'set independent torrent output dir'],
    ]:
        p.add_argument(*a[0], **{
            'default': a[1], 'action': 'store',
            'metavar': a[2], 'help': a[3]
        })
    p.add_argument('-T', '--threads', default=max_threads, action='store',
                   dest='max_threads', type=int, metavar='THREADS',
                   help='set number of threads THREADS (def: %s)' % max_threads)
    for enc_opt in enc_opts.keys():
        p.add_argument("--" + enc_opt, action=EncoderArg, nargs=0,
                       help='convert to %s' % (enc_opt))
    p.add_argument('flacdirs', nargs='+', metavar='flacdir',
                   help='directories to transcode')
    return p

def system(cmd):
    return os.system(cmd)

def transcode(f, flacdir, mp3_dir, codec, opts, lock):
    tags = {}
    for tag in copy_tags:
        tagcmd = "metaflac --show-tag='" + escape_quote(tag) + \
                 "' '" + escape_quote(f) + "'"
        t = re.sub(r'\S.+?=', '', os.popen(tagcmd).read().rstrip(), count=1)
        if t:
            tags.update({tag: escape_quote(t)})
        del t
    if (opts.zeropad and 'TRACKNUMBER' in tags
       and len(tags['TRACKNUMBER']) == 1):
        tags['TRACKNUMBER'] = '0' + tags['TRACKNUMBER']
    if opts.skipgenre and 'GENRE' in tags: del tags['GENRE']

    outname = re.sub(re.escape(flacdir), mp3_dir, f)
    outname = re.sub(re.compile(r'\.flac$', re.IGNORECASE), '', outname)
    with lock:
        if not os.path.exists(os.path.dirname(outname)):
            os.makedirs(os.path.dirname(outname))
    outname += enc_opts[codec]['ext']
    if os.path.exists(outname) and not opts.overwrite:
        print("WARN: file %s already exists" % (os.path.relpath(outname)),
              file=sys.stderr)
        return 1
    flac_cmd = encoders[enc_opts[codec]['enc']]['enc']
    tagline = ''
    for tag in tags:
        tagline = tagline + " " + encoders[enc_opts[codec]['enc']][tag]
    tagline = tagline % tags
    if opts.dither:
        flac_cmd = dither_cmd + ' | ' + flac_cmd
    flac_cmd = "flac -sdc -- '" + escape_percent(escape_quote(f)) + \
               "' | " + flac_cmd
    flac_cmd = flac_cmd % {
        'opts': enc_opts[codec]['opts'],
        'filename': escape_quote(outname),
        'tags': tagline
    }
    outname = os.path.basename(outname)
    if not opts.silent:
        print("encoding %s" % outname)
    if opts.verbose:
        print(flac_cmd)
    r = system(flac_cmd)
    if r:
        failure(r, "error encoding %s" % outname)
        system("touch '%s/FAILURE'" % mp3_dir)
    return 0

class Transcode(threading.Thread):
    def __init__(self, file, flacdir, mp3_dir, codec, opts, lock, cv):
        threading.Thread.__init__(self)
        self.file = file
        self.flacdir = flacdir
        self.mp3_dir = mp3_dir
        self.codec = codec
        self.opts = opts
        self.lock = lock
        self.cv = cv

    def run(self):
        r = transcode(self.file, self.flacdir, self.mp3_dir, self.codec,
                      self.opts, self.lock)
        with self.cv:
            self.cv.notify_all()
        return r

def main():
    parser = setup_parser()
    opts = parser.parse_args()
    if not opts.output.endswith('/'):
        opts.output += '/'
    if len(codecs) == 0 and not opts.original:
        parser.error("you must provide at least one format to transcode to")
        exit()
    for flacdir in opts.flacdirs:
        flacdir = os.path.abspath(flacdir)
        flacfiles = []
        if not os.path.exists(opts.torrent_dir):
            os.makedirs(opts.torrent_dir)
        for dirpath, dirs, files in os.walk(flacdir, topdown=False, followlinks=True):
            for name in files:
                if fnmatch(name.lower(), '*.flac'):
                    flacfiles.append(os.path.join(dirpath, name))
        flacfiles.sort()
        if opts.ignore and not flacfiles:
            if not opts.silent:
                print("SKIP (no flacs in): %s" % (os.path.relpath(flacdir)))
            continue
        if opts.original:
            if not opts.silent:
                print('BEGIN ORIGINAL FLAC')
            if opts.output and opts.tracker and not opts.notorrent:
                make_torrent(opts, flacdir)
            if not opts.silent:
                print('END ORIGINAL FLAC')

        for codec in codecs:
            outdir = os.path.basename(flacdir)
            flacre_str = '((16|24)+(-| )*(bit|44|48|96)* *)*FLAC( *(16|24)+(-| )*(bit|44|48|96)*)*'
            flacre = re.compile(flacre_str, re.IGNORECASE)
            if flacre.search(outdir):
                outdir = flacre.sub(codec, outdir)
            else:
                outdir = outdir + " (" + codec + ")"
            outdir = opts.output + outdir
            if not os.path.exists(outdir):
                os.makedirs(outdir)

            if not opts.silent:
                print('BEGIN ' + codec + ': %s' % os.path.relpath(flacdir))
            threads = []
            cv = threading.Condition()
            lock = threading.Lock()
            for f in flacfiles:
                with cv:
                    while (threading.active_count() == max(1, opts.max_threads) + 1):
                        cv.wait()
                    t = Transcode(f, flacdir, outdir, codec, opts, lock, cv)
                t.start()
                threads.append(t)
            for t in threads:
                t.join()

            if not opts.nocopyother:
                copy_other(opts, flacdir, outdir)
            if opts.replaygain:
                replaygain(opts, codec, outdir)
            if opts.output and opts.tracker and not opts.notorrent:
                make_torrent(opts, outdir)
            if not opts.silent:
                print('END ' + codec + ': %s' % os.path.relpath(flacdir))

        if opts.verbose: print('ALL DONE: ' + os.path.relpath(flacdir))
    return 0

if __name__ == '__main__':
    main()
